DataFormats/Portable/interface/PortableDeviceCollection.h

```diff
@@ -8,37 +8,115 @@
 #include "HeterogeneousCore/AlpakaInterface/interface/config.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/traits.h"
+#include "DataFormats/Portable/interface/PortableCollectionCommon.h"
 
 // generic SoA-based product in device memory
-template <typename T, typename TDev, typename = std::enable_if_t<cms::alpakatools::is_device_v<TDev>>>
+template <typename TDev,
+          typename T0,
+          typename T1 = void,
+          typename T2 = void,
+          typename T3 = void,
+          typename T4 = void,
+          typename = std::enable_if_t<cms::alpakatools::is_device_v<TDev>>>
 class PortableDeviceCollection {
   static_assert(not std::is_same_v<TDev, alpaka_common::DevHost>,
                 "Use PortableHostCollection<T> instead of PortableDeviceCollection<T, DevHost>");
+  // Make sure void is not interleaved with other types.
+  static_assert(not std::is_same<T3, void>::value or std::is_same<T4, void>::value);
+  static_assert(not std::is_same<T2, void>::value or std::is_same<T3, void>::value);
+  static_assert(not std::is_same<T1, void>::value or std::is_same<T2, void>::value);
+
+  template <typename T>
+  static constexpr std::size_t count_t_ = portablecollection::typeCount<T, T0, T1, T2, T3, T4>;
+
+  template <typename T>
+  static constexpr std::size_t index_t_ = portablecollection::typeIndex<T, T0, T1, T2, T3, T4>;
+
+  static constexpr std::size_t members_ = portablecollection::membersCount<T0, T1, T2, T3, T4>;
 
 public:
-  using Layout = T;
-  using View = typename Layout::View;
-  using ConstView = typename Layout::ConstView;
   using Buffer = cms::alpakatools::device_buffer<TDev, std::byte[]>;
   using ConstBuffer = cms::alpakatools::const_device_buffer<TDev, std::byte[]>;
+  using Implementation = portablecollection::CollectionImpl<0, T0, T1, T2, T3, T4>;
+
+  using SizesArray = std::array<int32_t, members_>;
+
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  using Layout = portablecollection::TypeResolver<Idx, T0, T1, T2, T3, T4>;
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  using View = typename Layout<Idx>::View;
+  template <std::size_t Idx = 0, typename = std::enable_if_t<(members_ > Idx)>>
+  using ConstView = typename Layout<Idx>::ConstView;
+
+private:
+  template <std::size_t Idx>
+  using Leaf = portablecollection::CollectionLeaf<Idx, Layout<Idx>>;
+
+  template <std::size_t Idx>
+  Leaf<Idx>& get() {
+    return static_cast<Leaf<Idx>&>(impl_);
+  }
+
+  template <std::size_t Idx>
+  Leaf<Idx> const& get() const {
+    return static_cast<Leaf<Idx> const&>(impl_);
+  }
+
+  template <typename T>
+  Leaf<index_t_<T>>& get() {
+    return static_cast<Leaf<index_t_<T>>&>(impl_);
```

Comment:
What happens if one tries to call `get()` with a type argument that occurs twice?

Reply:
IIUC, that would result in a compilation error.